python/mspasspy/algorithms/basic.py

```diff
@@ -1,5 +1,55 @@
 from mspasspy.util.decorators import mspass_func_wrapper
 from mspasspy.ccore.algorithms.basic import LinearTaper, CosineTaper, VectorTaper
+from mspasspy.ccore.utility import MsPASSError, ErrorSeverity
+from mspasspy.util import logging_helper
+import mspasspy.ccore.algorithms.basic as bsc
+from mspasspy.ccore.seismic import (
+    TimeSeries,
+    TimeSeriesEnsemble,
+    Seismogram,
+    SeismogramEnsemble,
+)
+
+
+@mspass_func_wrapper
+def ExtractComponent(
+    data,
+    component,
+    *args,
+    object_history=False,
+    alg_name=None,
+    alg_id=None,
+    dryrun=False,
+    inplace_return=False,
+    function_return_key=None,
+    **kwargs
+):
+    if isinstance(data, Seismogram):
+        empty = TimeSeries()
```

Comment:
It is not a good idea to create an empty data object here. We only need it when the input is dead or exception occurs. Otherwise, this is only wasting CPU cycles and memory.

Also, the returned data object (whether being empty or not) should inherit the elog and object level history of the input data object. Note that the object level history is like a chain. This step converts a Seismogram into TimeSeries (or the corresponding Ensemble type), and should be considered a part of the chain.

Reply:
I will add load_history() here to copy the ProcessingHistory of input data to the new TimeSeries object